lib/errors/index.js

```diff
@@ -226,21 +226,96 @@ exports.UnknownConstraintError = UnknownConstraintError;
+ *
  * @param {string} value The value that generated the error
  */
 class ValidationErrorItem {
-  constructor(message, type, path, value) {
+  constructor(message, type, path, value, inst, validatorKey, fnName, fnArgs) {
     this.message = message || '';
-    this.type = type || null;
+    this.type = null;
     this.path = path || null;
     this.value = value !== undefined ? value : null;
-    //This doesn't need captureStackTrace because its not a subclass of Error
+
+    this.origin         = null;
+    this.instance       = inst || null;
+    this.validatorKey   = validatorKey || null;
+    this.validatorName  = fnName || null;
+    this.validatorArgs  = fnArgs || [];
+
+    if (type) {
+      if (ValidationErrorItem.ORIGINS[ type ]) {
+        this.origin = type;
+      } else {
+        const lowercaseType = (type + '').toLowerCase().trim();
+        const realType  = ValidationErrorItem.TypeStringMap[ lowercaseType ];
+
+        if (realType && ValidationErrorItem.ORIGINS[ realType ]) {
+          this.origin = realType;
+          this.type = type;
+        }
+      }
+    }
+  }
+
+  /**
+   * return a lowercase, trimmed string "key" that identifies the validator.
+   *
+   * N.B. the string will be empty if the instance has neither a valid `validatorKey` property nor a valid `validatorName` property
+   *
+   * @param   {Boolean} useTypeAsNS   - optional, defaults to TRUE, controls whether the returned value is "namespace",
+   *                                    this parameter is ignored if the validator's `type` is not one of ValidationErrorItem.ORIGINS
+   * @param   {String}  NSSeparator   - a separator string for concatenating the namespace, must be not be empty,
+   *                                    defaults to "." (fullstop). only used and validated if useTypeAsNS is TRUE.
+   * @throws  {Error}                 - thrown if NSSeparator is found to be invalid.
+   * @return  {String}
+   */
+  getValidatorKey(useTypeAsNS/* = true */, NSSeparator/* = '.'*/) {
+    // parameter defaults: some CI instances give a syntax error on actual parameter default values
+    const
+      useTANS = typeof useTypeAsNS === 'undefined' ?  true : !!useTypeAsNS,
+      NSSep   = typeof NSSeparator === 'undefined' ? '.' : NSSeparator;
+
+    const
+      type  = this.origin,
+      key   = this.validatorKey || this.validatorName,
+      useNS = useTANS && type && ValidationErrorItem.ORIGINS[ type ];
+
+    if (useNS && (typeof NSSep !== 'string' || !NSSep.length)) {
+      throw new Error('Invalid namespace separator given, must be a non-empty string');
+    }
+
+    if (!(typeof key === 'string' && key.length)) {
+      return '';
+    }
+
+    return (useNS ? [type, key].join(NSSep) : key).trim().toLowerCase();
   }
 }
-exports.ValidationErrorItem = ValidationErrorItem;
+exports.ValidationErrorItem       = ValidationErrorItem;
+
+ValidationErrorItem.ORIGINS         = Object.freeze({
```

Comment:
I know right :-) ... alot of stuff that could/should be immutable isn't ... IMO it ought to be JS best practice to use immutability where ever possible, and this seems like a good candidate (additionally it's a call that is done only once when the lib is required so there is no appreciable overhead in the call to `Object.freeze()` (if it was being done in a place that could affect, for instance, every DAO instance then obviously the possible performance ramifications would weigh alot more heavily)

but if it's keep you from signing off on this PR then just say the word and I'll remove the `Object.freeze()` call